dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/monitor/DatadogRumMonitor.kt

```diff
@@ -0,0 +1,165 @@
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-2020 Datadog, Inc.
+ */
+
+package com.datadog.android.rum.internal.monitor
+
+import com.datadog.android.core.internal.data.Writer
+import com.datadog.android.core.internal.time.TimeProvider
+import com.datadog.android.core.internal.utils.devLogger
+import com.datadog.android.rum.GlobalRumMonitor
+import com.datadog.android.rum.RumMonitor
+import com.datadog.android.rum.internal.domain.RumEvent
+import com.datadog.android.rum.internal.domain.RumEventData
+import java.util.UUID
+
+internal class DatadogRumMonitor(
+    private val timeProvider: TimeProvider,
+    private val writer: Writer<RumEvent>
+) : RumMonitor {
+
+    private val activeView = mutableMapOf<Any, RumEvent>()
+    private val activeResources = mutableMapOf<Any, RumEvent>()
+
+    // region RumMonitor
+
+    override fun addUserAction(
+        action: String,
+        attributes: Map<String, Any?>
+    ) {
+        val eventData = RumEventData.UserAction(action)
+        val event = RumEvent(
+            GlobalRumMonitor.getRumContext(),
+            timeProvider.getServerTimestamp(),
+            eventData,
+            attributes
+        )
+        writer.write(event)
+    }
+
+    override fun startView(
+        key: Any,
+        name: String,
+        attributes: Map<String, Any?>
+    ) {
+        val activeViewId = UUID.randomUUID()
+        GlobalRumMonitor.updateViewId(activeViewId)
+        val eventData = RumEventData.View(name, System.nanoTime())
+        val event = RumEvent(
+            GlobalRumMonitor.getRumContext(),
+            timeProvider.getServerTimestamp(),
+            eventData,
+            attributes
+        )
+        activeView[key] = event
+    }
+
+    override fun stopView(
+        key: Any,
+        className: String,
+        attributes: Map<String, Any?>
+    ) {
+        val startedEvent = activeView.remove(key)
+        val startedEventData = startedEvent?.eventData as? RumEventData.View
+
+        if (startedEvent == null) {
+            devLogger.w(
+                "Unable to stop view with key <$key> and class <$className>. " +
+                    "This can mean that the view was not started or already stopped."
+            )
+        } else if (startedEventData == null) {
+            devLogger.e(
+                "Unable to stop view with key <$key> and class <$className>. " +
+                    "The related data was inconsistent."
+            )
+        } else {
+            val updatedEvent = startedEvent.copy(
+                eventData = startedEventData.copy(
+                    durationNanoSeconds = System.nanoTime() - startedEventData.durationNanoSeconds,
+                    version = startedEventData.version + 1
+                ),
+                attributes = startedEvent.attributes.toMutableMap().apply {
```

Comment:
Do you need to do this ? Isn't the copy doing this by default ? (taking the attributes from the prev object)